Add support for Apollo Connectors Mapping Syntax

phryneas opened the pull request.

phryneas committed:
minimal changes to inject new grammar into directives

phryneas committed:
minimal (wrong) syntax highlighting for body & selection

phryneas commented on syntaxes/graphql.connectors.json:
```diff
@@ -0,0 +1,105 @@
+            "3": { "name": "invalid.illegal.newline.graphql" }
+          },
+          "patterns": [
+            { "include": "source.graphql#graphql-field" },
+            { "include": "source.graphql#graphql-comma" }
```
This is the point to specify extra rules for the real grammar

phryneas committed:
body...

phryneas committed:
first attempt at jsonselection textmate grammar

phryneas committed:
injection not working yet

phryneas commented on syntaxes/graphql.connectors.json:
```diff
@@ -0,0 +1,68 @@
+      },
+      "patterns": [
+        {
+          "contentName": "meta.embedded.block.jsonselection.graphql",
```
Right now I'm stuck at the point where I'd need to inject the JSONSelection grammar into here.

phryneas committed:
embed jsonselection as language

phryneas committed:
some minor highlighting

phryneas committed:
change to new ebnf, start over

phryneas committed:
progress

phryneas committed:
progress

phryneas committed:
getting somewhere

phryneas committed:
JSONSelectionString

phryneas committed:
add comments, snapshot tests

phryneas committed:
update one grammar, move things

phryneas committed:
uncomment

dylan-apollo commented on syntaxes/connectors.jsonselection.yaml:
```diff
@@ -0,0 +1,270 @@
+# use 'js-yaml' to convert to json
+# npx js-yaml syntaxes/connectors.jsonselection.yaml >| syntaxes/connectors.jsonselection.json
+name: "Apollo Connectors JSONSelection"
```
Is this user-facing at all? We've avoided using this name anywhere user-facing for now, since we're going to change it, we just haven't decided to what yet 😅

phryneas replied:
That's a good call-out - the only "user-facing" thing right now is that we introdude a new language to VSCode:

https://github.com/apollographql/vscode-graphql/blob/0ea12de909a837847cdedc704c80f4278f3ec879/package.json#L159-L165

That means it will pop up if you edit a file called `test.apollo.jsonselection`, but also if you use the "change file language" dropdown:

![screenshot of that dropdown](image.png)

We might want to give that language a better user-facing name.

It's technically also visible if you inspect syntax highlighting tokens, but I would not really consider that "user-facing":

![screenshot](image.png)

dylan-apollo replied:
Yeah, that sounds generic enough to be okay. We also use the term "mapping" a lot when referring to it right now, like "Apollo Connectors Mapping Syntax", though that's really verbose.

phryneas replied:
![image](https://github.com/user-attachments/assets/74f57c64-12a0-4858-b394-56363ec5318a)

Renamed it, and also put it into `[]` because that will sort it alphabetically at the bottom of the list. There's no need for this implementation detail to be the first thing on the list every time people open it :)

dylan-apollo commented on syntaxes/connectors.jsonselection.yaml:
```diff
@@ -0,0 +1,270 @@
+  NamedPathSelection:
+    begin: "\\b([a-zA-Z][0-9a-zA-Z_]*)\\s*(:)"
+    beginCaptures:
+      "1":
+        name: "string.unquoted.NamedPathSelection.alias"
```
So can each of these sections get their own colors? I'd love to have one color for the "output" name (alias in this case) and another for things that are inputs.

phryneas replied:
Yes, they can!

Open a text editor and execute the "Developer: inspect editor tokens and scopes" command

![image](https://github.com/user-attachments/assets/9c11445e-65cc-4b8a-ae05-742de22f9da1)

that will give you this pop-up:

![image](https://github.com/user-attachments/assets/c03c1339-1d9a-4d6d-9695-aed9542d1df2)


And then you can move your cursor around and see what tokens things in other languages have, as an inspiration.

---

![image](https://github.com/user-attachments/assets/c3af9462-8e1a-45e6-ac59-84d782649d49)
![image](https://github.com/user-attachments/assets/b8403166-dc27-438d-86a4-51e2da416e18)

You see that GraphQL uses `string.unquoted` and `variable` as highlighting scopes here, but I also see that the `alias` is never applied because `variable.KeyPath.key.identifier` seems to have a higher priority somehow.

I'll debug that :)

phryneas replied:
So... with some changes:

![image](https://github.com/user-attachments/assets/2f83e00b-f560-455c-b51f-5c7a79c3a61a)

dylan-apollo replied:
Amazing! A couple more notes on what we might want to add in future iterations:

1. `$` and `@` as well as an identifier following $, like `$args`, `$this`, etc. are sort of our keywords? They're special, at very least, and might be worth highlighting separately.
2. Anything in `$()` or as an argument to a method is sort of a different language mode. We call them literals, but they can also have some expressions within them. So `${ blah: $.blah }` and `$({ blah: $.blah })` are both valid, and look very similar, but do different things. It'd be nice to highlight that difference somehow (effectively, just the presence of the parens causes it).

phryneas replied:
I can configure `{(` and `)}` as a special kind of bracket pairs (this is VSCode, not TextMate!) so they will always have the same color and highlight in pairs together:
![image](https://github.com/user-attachments/assets/c6d49827-a451-410b-9da8-b9175f972975)

lennyburdette replied:
that's cool! i don't think it's exactly what we want tho. the way i see it, when we're inside parens (for both literal values like `$(...)` or for method arguments like `->method(...)`), we're now in "literal value" mode instead of "selection" mode. So I'd color the outer parens as one thing, and any inner values to indicate that they're different from selections. i dunno exactly what that looks like

phryneas replied:
As for 1., does this work for you @dylan-apollo ?
![image](https://github.com/user-attachments/assets/9fb17a95-52af-4178-abb7-362d45115ad9)

phryneas replied:
@lennyburdette then I'm undoing that change and we go back to this:
![image](https://github.com/user-attachments/assets/97f50d89-1740-4547-b840-cc40bc2dede6)


Honestly I don't know if I can do a lot to visually highlight that difference more, apart from adding a full other "sub-language", which I'd rather avoid as this would become even more unmaintainable 😅

phryneas replied:
> Maybe we just color parens differently than {} for now, since parens always represent entering literal mode? We could do something special with semantic tokens or something if we really want in the future

We'll need to test that out in the non-VSCode editors. VSCode pretty much ignores everything else and applies its "bracket pair coloring" logic over the top of configured bracket pairs (which we want to have so it highlights start and end).

dylan-apollo replied:
Then maybe it's just fine 😅. Certainly shouldn't block getting the highlighting of the rest of the language out there

dylan-apollo commented on syntaxes/graphql.connectors.json:
```diff
@@ -0,0 +1,108 @@
+    },
+    "connect-jsonselection-argument": {
+      "name": "meta.argument.connect.jsonselection.graphql",
+      "begin": "\\s*(selection|body)(?:\\s*(:))",
```
Are these recursive or only top-level? `body` is within `http`, is the only reason I ask

phryneas replied:
Semi-rescursive. If you look a few lines up you'll see that I have an inlined pattern in `#connect-arguments` that looks for `\\s*(http)(?:\\s*(:))` and then applies `#connect-arguments` to that recursively.

Although I think this might actually work more on accident... I'll give this another read or two :D

phryneas replied:
Should be more solid now, I added a `#connect-jsonselection-nested-argument` pattern

dylan-apollo commented on syntaxes/graphql.connectors.json:
```diff
@@ -0,0 +1,108 @@
+    "connect-string-jsonselection-argument": {
+      "name": "meta.argument.connect.string.jsonselection.graphql",
+      "contentName": "meta.embedded.block.string.jsonselection.graphql",
+      "begin": "\\s*(GET|POST)(?:\\s*(:))",
```
Oh I see, is this what is checking the sub-argument to find `body`? We also support `PUT`, `PATCH`, and `DELETE`

phryneas replied:
This is to find usages where we don't do a full jsonselection "selection set", but a string that contains selections between `{}`.

phryneas replied:
Also, added support for PUT|PATCH|DELETE ✅

dylan-apollo reviewed (approved):
I'll be honest, I don't really understand how this works, but your demo looked great! Added a few questions/comments

phryneas committed:
flip some priorities

phryneas committed:
allow leading underscore in identifiers/keys/aliases

phryneas committed:
allow PUT|PATCH|DELETE

phryneas committed:
split out connect-jsonselection-nested-argument, test

phryneas committed:
add schema ref

phryneas committed:
fix: correct handling of nested objects
```graphql
type Mutation {
  createListing: String
    @connect(
      http: {
        body: """
        foo: bar
        """
      }
    )
}

type Query {
  user(id: ID!): User
}
```

phryneas committed:
special highlighting for `({` `})` bracket pairs

phryneas committed:
fix typo

phryneas committed:
highlighting of `@`, `$.` and `$.foo`

phryneas committed:
undo `{(` grouping

phryneas committed:
set apart `{` in `JSONSelectionString` visually

phryneas committed:
redo accidental wrong edit

phryneas committed:
update snapshots, run more tests in CI

phryneas committed:
also write graphql.connectors with a yaml source

phryneas committed:
rename scope according to recommendations

phryneas committed:
rename things to "Apollo Connectors Mapping Syntax"

phryneas committed:
sort the language at the end of the list

phryneas committed:
reduce diff to original graphql grammar

phryneas committed:
add comments to document injection logic

phryneas commented on syntaxes/graphql.json:
```diff
@@ -342,7 +342,14 @@
     },
     "graphql-directive": {
-      "begin": "\\s*((@)\\s*([_A-Za-z][_0-9A-Za-z]*))",
+      "contentName": "meta.embedded.block.directive.graphql",
+      "begin": "\\s*(?=@)",
+      "end": "(?=.)",
+      "applyEndPatternLast": 1,
+      "patterns": [{ "include": "#graphql-directive-body" }]
+    },
+    "graphql-directive-body": {
+      "begin": "((@)\\s*([_A-Za-z][_0-9A-Za-z]*))",
```
I managed to get this down to a minimum amount of changes, but unfortunately not down to none.

phryneas commented on syntaxes/graphql.connectors.yaml:
```diff
@@ -0,0 +1,167 @@
+scopeName: source.apollo.connectors.graphql
+# inject these rules into source.graphql#graphql-directive, before all other patterns (:L), ignoring strings and comments
+injectionSelector: L:meta.embedded.block.directive.graphql -string -comment
+patterns:
+  - include: "#connect-directive"
+repository:
+  # if we encounter a `@connect` directive, move forward into this set of rules, otherwise it will just follow the default graphql rules
```
I added a good bunch of comments to this file, please let me know if they are helpful or if I could make something more clear here.

phryneas commented on .vscode/settings.json:
```diff
@@ -12,5 +14,8 @@
-  "debug.node.autoAttach": "on"
+  "debug.node.autoAttach": "on",
+  "yaml.schemas": {
+    "https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json": "/syntaxes/*.yaml"
```
I found this JSONSchema which *somehow* adds a tiny little bit of documentation to the textmate format.

phryneas commented on src/language-server/__tests__/fixtures/documents/functionCall.ts.snap:
```diff
@@ -413,11 +413,11 @@
->  `query { test }`
+>  `query Q11 { test }`
 #^^ source.ts
 #  ^ source.ts string.template.ts punctuation.definition.string.template.begin.ts
-#   ^^^^^^^^^^^^^^ source.ts string.template.ts
-#                 ^ source.ts string.template.ts punctuation.definition.string.template.end.ts
+#   ^^^^^^^^^^^^^^^^^^ source.ts string.template.ts
+#                     ^ source.ts string.template.ts punctuation.definition.string.template.end.ts
```
Unrelated, it seems I haven't updated some test snapshots in a while.

phryneas commented:
I think this is in a good state right now, so I'll wait a bit for feedback from testing before we can hopefully get this in next week or so :)

dylan-apollo commented:
@phryneas this language is going to show up in a few more places now, I'm just wondering how hard it will be to extend syntax highlighting to these as well?

1. In both `@source` and `@connect`, the `http.headers` value is a list of objects or a single object. If an object it has a `value` property (string), which contains `{ }`, then the content within those `{ }` is this same language.
2. Very similar to 1, but _only_ for `@connect`, the string of `http.GET`, `http.POST`, `http.PUT`, `http.PATCH`, and `http.DELETE` can contain expressions within `{ }`.

I believe @BoD said we can use semantic tokens (or whatever they're called) via LSP for JetBrains now, so maybe this is complex enough that we should look into that route?

phryneas committed:
Merge remote-tracking branch 'origin/main' into pr/textmate

phryneas commented:
> @phryneas this language is going to show up in a few more places now, I'm just wondering how hard it will be to extend syntax highlighting to these as well?
> 
>     1. In both `@source` and `@connect`, the `http.headers` value is a list of objects or a single object. If an object it has a `value` property (string), which contains `{ }`, then the content within those `{ }` is this same language.
> 
>     2. Very similar to 1, but _only_ for `@connect`, the string of `http.GET`, `http.POST`, `http.PUT`, `http.PATCH`, and `http.DELETE` can contain expressions within `{ }`.
> 

That would be a duplication of most of `syntaxes/graphql.connectors.json`, with some adjustments. Probably not too complex.

> 
> I believe @BoD said we can use semantic tokens (or whatever they're called) via LSP for JetBrains now, so maybe this is complex enough that we should look into that route?

Semantic tokens won't work for the documentation, and they only work once rover is installed & active, while this approach would always work.
Usually, semantic tokens are used to refine the highlighting of a textmate grammar, not to completely replace it, as afaik it can also be a lot slower and might lag behind while typing - but in the end, it's your call to make.

phryneas committed:
add accurate link

phryneas committed:
add build script, extract common variables

phryneas committed:
WIP

phryneas committed:
more build logging

phryneas committed:
add "debug" mode to build step, remove hand-debugging names from grammar

phryneas committed:
wip

phryneas committed:
progress

phryneas committed:
LitExpr mostly done

phryneas committed:
NamedSelection

phryneas committed:
Merge commit 'dc6b7bb5c10272a506edb867a3f018b7b40193d3' as 'sampleWorkspace/connectors-community'

phryneas committed:
Squashed 'sampleWorkspace/connectors-community/' content from commit 31a22035

git-subtree-dir: sampleWorkspace/connectors-community
git-subtree-split: 31a2203549f4879b7e0ca22345d9cda78a5ab919

phryneas committed:
add connectors to sampleWorkspace

phryneas committed:
add command to update subtree, semgrep

phryneas committed:
add back `NamedFieldSelection`

phryneas committed:
extract styling

phryneas committed:
adjust semgrepignore

phryneas committed:
fix up cross-referencing styles

phryneas committed:
test1

phryneas committed:
embed test

phryneas committed:
allow comments

phryneas committed:
add circleCI run

phryneas committed:
re-add working dir

phryneas committed:
add job to workflow

phryneas committed:
allow for single-line selection string

phryneas committed:
more one-line string handling

phryneas committed:
CI

phryneas committed:
rename file

phryneas committed:
remove "Run other tests" step

phryneas committed:
basic addition for headers

phryneas committed:
apply to `@connect` and `@source`

phryneas committed:
multiline string support

phryneas committed:
highlight string contents

phryneas committed:
highlight known builtin functions

phryneas committed:
change property/aliases to be more in line with graphql highlighting

phryneas committed:
Merge commit 'a52c6b4b839001d2c18e63080e41193539dbd4d6' into pr/textmate

phryneas committed:
Squashed 'sampleWorkspace/connectors-community/' changes from 31a22035..2c7ee28f

2c7ee28f fix formatting that breaks syntax highlighting (#69)

git-subtree-dir: sampleWorkspace/connectors-community
git-subtree-split: 2c7ee28fc66a882f82bf7d3f2aea0a32eda35fba

phryneas commented:
Okay, finalizing the colors. Here are a few impressions:


<img width="932" height="1101" alt="image" src="https://github.com/user-attachments/assets/7ef58d88-5a76-4a3a-ae9b-00e184e7e8cb" />


Here you can see that the object syntax (top) is different from the selection syntax (bottom).
This mimicks the colors that GraphQL syntax is using:

<img width="207" height="257" alt="image" src="https://github.com/user-attachments/assets/3c37fcd7-4063-48b6-abc1-fd1eb32376f9" />

Another thing to point out is that builtin functions are highlighted differently:

<img width="498" height="141" alt="image" src="https://github.com/user-attachments/assets/2908fdd5-04f8-46b4-9f07-41df62cd7d95" />

The colors look very different here, but that's theme-specific. I'm using `entity.name.function` for a random function name and `keyword.builtin, entity.name.function` for known function names.

phryneas committed:
highlight embedded quotes

phryneas committed:
rename all jsonselection to connectors.mapping

phryneas committed:
"change language" in embedded selection string

phryneas committed:
add `oniguruma-parser/optimizer` to build step

phryneas committed:
add option

dylan-apollo reviewed (approved):
I don't know much about the implementation, but I'm thrilled to have syntax highlighting!

My only note stylistically is I feel like the identifier on the left side of a `:` and the value on the right should have their colors swapped. In the screenshot, it makes sense to be that the short form is the same as the property side of objects (in `headers`). But in the long form, it seems like the key side should match the keys in objects and the value side should match the literal value `$("blah")`.

If that doesn't make sense, happy to huddle on it!

<img width="1328" height="480" alt="Image" src="https://github.com/user-attachments/assets/f413c48e-f4a9-4ff1-ada5-f833b77a51d2" />

<img width="522" height="140" alt="Image" src="https://github.com/user-attachments/assets/b72e38e1-9eb2-4c58-8f47-e64158f0e006" />

phryneas committed:
review session with Dylan

phryneas committed:
add changeset

phryneas committed:
Merge branch 'main' into pr/textmate

chaiphet6669-afk reviewed (commented):
`````# - @wtgtybhertgeghgtwtg [__](**increment the:

MAJOR version when you make incompatible API changes MINOR version when you add functionality in a backward compatible manner PATCH version when you make backward compatible bug fixes 

**Example:**

_git tag -s 1.0.1 -m "Bump version: 1.0.0 to 1.0.1_

**)````

- [x] - ```. ![0 comman.txt](https://github.com/user-attachments/files/30298655/0.comman.txt)

chaiphet6669-afk reviewed (commented):
sampleWorkspace/connectors-community/.vscode/tasks.json

chaiphet6669-afk reviewed (commented):
- - [ ] - - [_sampleWorkspace/connectors-community/README.md_ ] -
